linkerd/app/src/trace_collector.rs

```diff
@@ -0,0 +1,114 @@
+use linkerd_app_core::http_tracing::{CollectorProtocol, SpanSink};
+use linkerd_app_core::metrics::ControlHttp as HttpMetrics;
+use linkerd_app_core::svc::NewService;
+use linkerd_app_core::{control, dns, identity, opencensus, opentelemetry};
+use linkerd_error::Error;
+use std::collections::HashMap;
+use std::future::Future;
+use std::pin::Pin;
+use tracing::info_span;
+
+pub mod oc_collector;
+pub mod otel_collector;
+
+const SPAN_BUFFER_CAPACITY: usize = 100;
+const SERVICE_NAME: &str = "linkerd-proxy";
+
+#[derive(Clone, Debug)]
+pub enum Config {
+    Disabled,
+    Enabled(Box<EnabledConfig>),
+}
+
+#[derive(Clone, Debug)]
+pub struct EnabledConfig {
+    pub control: control::Config,
+    pub attributes: HashMap<String, String>,
+    pub hostname: Option<String>,
+    pub kind: CollectorProtocol,
+}
+
+pub type Task = Pin<Box<dyn Future<Output = ()> + Send + 'static>>;
+
+pub enum TraceCollector {
+    Disabled,
+    Enabled(Box<EnabledCollector>),
+}
+
+pub struct EnabledCollector {
+    pub addr: control::ControlAddr,
+    pub kind: CollectorProtocol,
+    pub span_sink: SpanSink,
+    pub task: Task,
+}
+
+impl TraceCollector {
+    pub fn span_sink(&self) -> Option<SpanSink> {
+        match self {
+            TraceCollector::Disabled => None,
+            TraceCollector::Enabled(inner) => Some(inner.span_sink.clone()),
+        }
+    }
+}
+
+impl Config {
+    pub fn info_span(&self) -> tracing::Span {
+        match self {
+            Config::Disabled => info_span!("disabled_tracing"),
+            Config::Enabled(config) => match config.kind {
+                CollectorProtocol::OpenCensus => info_span!("opencensus"),
+                CollectorProtocol::OpenTelemetry => info_span!("opentelemetry"),
+            },
+        }
+    }
+
+    pub fn metrics_prefix(&self) -> &'static str {
+        match self {
+            Config::Disabled => "disabled_tracing",
```

Comment:
Does this mean we're going to start exporting a bunch of metrics with the disabled_tracing_... in the default case? That's pretty yucky.

It would be better, imo, to stub out the metrics registry when tracing is disabled so that we don't add any metrics to the global registry.

Reply:
Done.